docs/architecture/adr-038-state-listening.md

```diff
@@ -182,349 +137,587 @@ func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
-func NewSimApp(
-	logger log.Logger,
-	db dbm.DB,
-	traceStore io.Writer,
-	loadLatest bool,
-	appOpts servertypes.AppOptions,
-	baseAppOptions ...func(*baseapp.BaseApp),
-) *SimApp {
+func NewStreamingPlugin(name string, logLevel string) (interface{}, error) {
+    logger := hclog.New(&hclog.LoggerOptions{
+       Output: hclog.DefaultOutput,
+       Level:  toHclogLevel(logLevel),
+       Name:   fmt.Sprintf("plugin.%s", name),
+    })
+
+    // We're a host. Start by launching the streaming process.
+    env := os.Getenv(GetPluginEnvKey(name))
+    client := plugin.NewClient(&plugin.ClientConfig{
+       HandshakeConfig: HandshakeMap[name],
+       Plugins:         PluginMap,
+       Cmd:             exec.Command("sh", "-c", env),
+       Logger:          logger,
+       AllowedProtocols: []plugin.Protocol{
+           plugin.ProtocolNetRPC, plugin.ProtocolGRPC},
+    })
+
+    // Connect via RPC
+    rpcClient, err := client.Client()
+    if err != nil {
+       return nil, err
+    }
 
-	...
-
-	keys := sdk.NewKVStoreKeys(
-	authtypes.StoreKey, banktypes.StoreKey, stakingtypes.StoreKey,
-	minttypes.StoreKey, distrtypes.StoreKey, slashingtypes.StoreKey,
-	govtypes.StoreKey, paramstypes.StoreKey, ibchost.StoreKey, upgradetypes.StoreKey,
-	evidencetypes.StoreKey, ibctransfertypes.StoreKey, capabilitytypes.StoreKey,
-	)
-
-	pluginsOnKey := fmt.Sprintf("%s.%s", plugin.PLUGINS_TOML_KEY, plugin.PLUGINS_ON_TOML_KEY)
-	if cast.ToBool(appOpts.Get(pluginsOnKey)) {
-		// this loads the preloaded and any plugins found in `plugins.dir`
-		pluginLoader, err := loader.NewPluginLoader(appOpts, logger)
-		if err != nil {
-			// handle error
-		}
-
-		// initialize the loaded plugins
-		if err := pluginLoader.Initialize(); err != nil {
-			// handle error
-		}
-
-		// register the plugin(s) with the BaseApp
-		if err := pluginLoader.Inject(bApp, appCodec, keys); err != nil {
-			// handle error
-		}
-
-		// start the plugin services, optionally use wg to synchronize shutdown using io.Closer
-		wg := new(sync.WaitGroup)
-		if err := pluginLoader.Start(wg); err != nil {
-			// handler error
-		}
-	}
+    // Request streaming plugin
+    return rpcClient.Dispense(name)
+}
 
-	...
+```
+
+We propose a `RegisterStreamingPlugin` function for the App to register `NewStreamingPlugin`s with the App's BaseApp.
+Streaming plugins can be of `Any` type; therefore, the function takes in an interface vs a concrete type.
+For example, we could have plugins of `ABCIListener`, `WasmListener` or `IBCListener`. Note that `RegisterStreamingPluing` function
+is helper function and not a requirement. Plugin registration can easily be moved from the App to the BaseApp directly.
 
-	return app
+```go
+// baseapp/streaming.go
+
+// RegisterStreamingPlugin registers the ABCI streaming service provided by the streaming plugin.
+func RegisterStreamingPlugin(
+    bApp *BaseApp,
+    appOpts servertypes.AppOptions,
+    keys map[string]*types.KVStoreKey,
+    streamingService interface{},
```

Comment:
```suggestion
    streamingPlugin interface{},
```

Reply:
Why not simply register `ABCIListener` here?
Can you elaborate on the "`WasmListener` or `IBCListener`", I guess to support those we'll need to modify cosmos-sdk anyway?

Reply:
>Why not simply register ABCIListener here?

To be able to support registering other types of plugins other than `ABCIListener`. 

Can you elaborate on the "WasmListener or IBCListener", I guess to support those we'll need to modify cosmos-sdk anyway?

They don't exist. They're examples of other services that can be exposed but there's no support for that.

As noted in the ADR, `RegisterStreamingPlugin` is an internal helper function not a requirement (which will most likely go away post merge. ref: https://github.com/cosmos/cosmos-sdk/pull/13473#discussion_r990118166)